test-tests/shared/src/test/scala/zio/test/BoolAlgebraSpec.scala

```diff
@@ -0,0 +1,168 @@
+          val expected = !failure1 && !failure2
+          assert(actual, equalTo(expected))
+        },
+        test("foreach combines multiple values") {
+          def isEven(n: Int): BoolAlgebra[String] =
+            if (n % 2 == 0) BoolAlgebra.success(s"$n is even")
+            else BoolAlgebra.failure(s"$n is odd")
+
+          val actual = BoolAlgebra.foreach(List(1, 2, 3))(isEven)
+          val expected = BoolAlgebra.failure("1 is odd") &&
+            BoolAlgebra.success("2 is even") &&
+            BoolAlgebra.failure("3 is odd")
+
+          assert(actual, isSome(equalTo(expected)))
+        },
+        test("implies returns implication of two values") {
+          assert(success1 ==> success2, isSuccess) &&
+          assert(success1 ==> failure1, isFailure) &&
+          assert(failure1 ==> success1, isSuccess) &&
+          assert(failure1 ==> failure2, isSuccess)
+        },
+        test("isFailure returns whether result is failure") {
+          assert(!success1.isFailure && failure1.isFailure, isTrue)
+        },
+        test("isSuccess returns whether result is success") {
+          assert(success1.isSuccess && !failure1.isSuccess, isTrue)
+        },
+        test("map transforms values") {
+          val actual   = (success1 && failure1 && failure2).map(_.split(" ").head)
+          val expected = BoolAlgebra.success("first") && BoolAlgebra.failure("first") && BoolAlgebra.failure("second")
+          assert(actual, equalTo(expected))
+        },
+        testM("or distributes over and") {
+          zio.test.check(boolAlgebra, boolAlgebra, boolAlgebra) { (a, b, c) =>
+            val left  = a || (b && c)
+            val right = (a || b) && (a || c)
+            assert(left, equalTo(right))
+          }
+        },
+        testM("or is associative") {
+          zio.test.check(boolAlgebra, boolAlgebra, boolAlgebra) { (a, b, c) =>
+            val left  = (a || b) || c
+            val right = a || (b || c)
+            assert(left, equalTo(right))
+          }
+        },
+        testM("or is commutative") {
+          zio.test.check(boolAlgebra, boolAlgebra) { (a, b) =>
+            assert(a || b, equalTo(b || a))
+          }
+        }
+      )
+    )
+
+object BoolAlgebraSpecHelper {
+  val value1 = "first success"
+  val value2 = "second success"
+  val value3 = "first failure"
+  val value4 = "second failure"
+
+  val success1 = BoolAlgebra.success(value1)
+  val success2 = BoolAlgebra.success(value2)
+  val failure1 = BoolAlgebra.failure(value3)
+  val failure2 = BoolAlgebra.failure(value4)
+
+  val isSuccess: Assertion[BoolAlgebra[_]] = assertion("isSuccess")()(_.isSuccess)
+  val isFailure: Assertion[BoolAlgebra[_]] = assertion("isFailure")()(_.isFailure)
+
+  def boolAlgebra: Gen[Random with Sized, BoolAlgebra[Int]] = Gen.small(s => boolAlgebraOfSize(s))
+
+  def boolAlgebraOfSize(size: Int): Gen[Random, BoolAlgebra[Int]] =
```

Comment:
I think right now this is generating objects that are "too large" which is why you are having to reduce the size and number of samples in checking the "hashCode is consistent with equals" property. I think we want the concept of "size" to be that each term adds one to the size, so for example `And(Not(Value(1)), Value(2))` would have a size of four, and then `boolAlgebraOfSize` should always generate objects of the requested size. I think something like the below will do that. When I tried that I was able to change the size and number of sample parameters back to their previous values.

```scala
  def boolAlgebra: Gen[Random with Sized, BoolAlgebra[Int]] = Gen.small(s => boolAlgebraOfSize(s), 1)

  def boolAlgebraOfSize(size: Int): Gen[Random, BoolAlgebra[Int]] =
    if (size == 1) {
      Gen.int(0, 9).map(BoolAlgebra.success)
    } else if (size == 2) {
      boolAlgebraOfSize(size - 1).map(!_)
    } else {
      for {
        n <- Gen.int(1, size - 2)
        gen <- Gen.oneOf(
                (boolAlgebraOfSize(n) <*> boolAlgebraOfSize(size - n - 1)).map((p => p._1 && p._2)),
                (boolAlgebraOfSize(n) <*> boolAlgebraOfSize(size - n - 1)).map(p => p._1 || p._2),
                boolAlgebraOfSize(size - 1).map(!_)
              )
      } yield gen
    }
```